http/fetch/archive_fetcher.go

```diff
@@ -36,19 +39,118 @@ import (
 	"github.com/fluxcd/pkg/tar"
 )
 
-// ArchiveFetcher holds the HTTP client that reties with back off when
-// the file server is offline.
+// ArchiveFetcher is a flexible API for downloading an archive from an HTTP server,
+// verifying its digest and extracting its contents to a given path in the Filesystem.
 type ArchiveFetcher struct {
-	httpClient        *retryablehttp.Client
+	retries           int
 	maxDownloadSize   int
 	maxUntarSize      int
+	fileMode          fs.FileMode
+	untar             bool
 	hostnameOverwrite string
+	filename          string
+	logger            any
+
+	httpClient *retryablehttp.Client
 }
 
+// Option is an option for constructing the ArchiveFetcher.
+type Option func(a *ArchiveFetcher)
+
 // ErrFileNotFound is an error type used to signal 404 HTTP status code responses.
 var ErrFileNotFound = errors.New("file not found")
 
+// WithRetries sets the maximum amount of retries the HTTP client will be allowed to make.
+func WithRetries(retries int) Option {
+	return func(a *ArchiveFetcher) {
+		a.retries = retries
+	}
+}
+
+// WithMaxDownloadSize specifies a limit for the size of the downloaded archive.
+func WithMaxDownloadSize(maxDownloadSize int) Option {
+	return func(a *ArchiveFetcher) {
+		a.maxDownloadSize = maxDownloadSize
+	}
+}
+
+// WithMaxUntarSize sets the option WithUntar(true) and specifies a limit for
+// the size of the extracted content of the tarball.
+func WithMaxUntarSize(maxUntarSize int) Option {
+	withUntar := WithUntar(true)
+	return func(a *ArchiveFetcher) {
+		withUntar(a)
+		a.maxUntarSize = maxUntarSize
+	}
+}
+
+// WithUntar tells the ArchiveFetcher to untar the tarball.
+func WithUntar(untar bool) Option {
+	return func(a *ArchiveFetcher) {
+		a.untar = untar
+	}
+}
+
+// WithHostnameOverwrite sets an override for the hostname in download URLs.
+func WithHostnameOverwrite(hostnameOverwrite string) Option {
+	return func(a *ArchiveFetcher) {
+		a.hostnameOverwrite = hostnameOverwrite
+	}
+}
+
+// WithLogger sets a logger for the HTTP client.
+// The logger can be any type that implements the retryablehttp.Logger or
+// retryablehttp.LeveledLogger interface. If the logger is of type logr.Logger,
+// it will be wrapped in a retryablehttp.LeveledLogger that only logs errors.
+func WithLogger(logger any) Option {
+	return func(a *ArchiveFetcher) {
+		a.logger = logger
+	}
+}
+
+// WithFileName sets the file name for the downloaded archive.
+func WithFileName(filename string) Option {
+	return func(a *ArchiveFetcher) {
+		a.filename = filename
+	}
+}
+
+// WithFileMode sets the file mode for the downloaded archive.
+// Applies only to unextracted archives.
```

Comment:
```suggestion
// Applies only if untar is enabled and the archive is extracted.
```

Reply:
No, that comment is wrong, this option only applies if untar is *not* enabled

Reply:
Maybe use that wording and just negate where needed 👍 Will do

Reply:
Yes, sorry about that. Ignore my comment.

Reply:
fixed, thanks